contracts/strategies/traderjoe/TraderJoeInvestmentLib.sol

```diff
@@ -0,0 +1,319 @@
+            amountXIn = depositTokenAmount;
+            amountYIn = pairDepositTokenAmount;
+        } else {
+            amountXIn = pairDepositTokenAmount;
+            amountYIn = depositTokenAmount;
+        }
+
+        uint256 binsAmount = strategyStorage.binIds.length;
+
+        (, , uint256 activeId) = strategyStorage.lbPair.getReservesAndId();
+
+        // Assume that token X and token Y have the same price.
+        uint256 totalAmount = amountXIn + amountYIn;
+        uint256 amountX;
+        uint256 amountY;
+
+        // Delta IDs.
+        int256[] memory deltaIds = new int256[](binsAmount);
+
+        // Distributions.
+        uint256[] memory distributionX = new uint256[](binsAmount);
+        uint256[] memory distributionY = new uint256[](binsAmount);
+
+        // The maximum of the number of bins is 51.
+        uint256 activeBinIndex = type(uint256).max;
+
+        for (uint256 i; i < binsAmount; ++i) {
+            deltaIds[i] = int256(strategyStorage.binIds[i]) - int256(activeId);
+
+            // Bin allocation has precision of 1e3.
+            uint256 amount = (totalAmount * strategyStorage.binAllocations[i]) /
+                1e3;
+
+            if (strategyStorage.binIds[i] < activeId) {
+                distributionY[i] = amount;
+                amountY += amount;
+            } else if (strategyStorage.binIds[i] > activeId) {
+                distributionX[i] = amount;
+                amountX += amount;
+            } else {
+                activeBinIndex = i;
+            }
+        }
+
+        // If one of our target bins is active, allocate rest of amountXIn and amountYIn to it to minimize swap fee.
+        if (activeBinIndex != type(uint256).max) {
+            uint256 amount = totalAmount - amountX - amountY;
+
+            if (amountXIn > amountX) {
+                uint256 amountXActive = Math.min(amountXIn - amountX, amount);
+
+                distributionX[activeBinIndex] = amountXActive;
+                amountX += amountXActive;
+
+                amount -= amountXActive;
+            }
+
+            if (amountYIn > amountY) {
+                uint256 amountYActive = Math.min(amountYIn - amountY, amount);
+
+                distributionY[activeBinIndex] = amountYActive;
+                amountY += amountYActive;
+
+                amount -= amountYActive;
+            }
+
+            if (amount > 0) {
+                revert InvalidActiveBinAllocation();
+            }
+        }
+
+        // Calibrate distributions so that the sum of them equals to 1e18, a precision of TraderJoe V2.
+        for (uint256 i; i < binsAmount; ++i) {
+            if (strategyStorage.binIds[i] < activeId) {
+                distributionY[i] = (distributionY[i] * 1e18) / amountY;
+            } else if (strategyStorage.binIds[i] > activeId) {
+                distributionX[i] = (distributionX[i] * 1e18) / amountX;
+            } else {
+                distributionX[i] = (distributionX[i] * 1e18) / amountX;
```

Comment:
in some corner cases amountX or amountY can be 0 which causes division by 0, right?

imagine an allocation when we deposit 100% into the active bin, but only supply amountXIn -> that case amountY is 0 right?

Reply:
No. The condition is not met.

Reply:
I think yes. 

If the allocation is 0,0,0,0,0,100%,0,0,0,0, and you only supply assetX to deposit, and the activeBin is the the bin where you are supposed to supply 100%, then this line never executes:
amountY += amountYActive;

which means this line is division by zero
  distributionY[i] = (distributionY[i] * 1e18) / amountY;
  
  no?

Reply:
I think yes.

If the allocation is 0,0,0,0,100%,0,0,0,0 and the amountYIn asset is not supplied to the deposit function and the active bin is the bin with 100% allocation, then the following line never executes:

amountY += amountYActive;

and this one also doesn't execute:

 amountY += amount;
 
 which means this will be division by zero:
 
 distributionY[i] = (distributionY[i] * 1e18) / amountY;
 
 no?

Reply:
Good spot, changed it. (Just for fyi, any allocation can't be 0.)